cardinal/router/router.go

```diff
@@ -0,0 +1,238 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"net"
+	"pkg.world.dev/world-engine/cardinal/txpool"
+	shardtypes "pkg.world.dev/world-engine/evm/x/shard/types"
+	routerv1 "pkg.world.dev/world-engine/rift/router/v1"
+	shard "pkg.world.dev/world-engine/rift/shard/v2"
+	"pkg.world.dev/world-engine/sign"
+)
+
+const (
+	defaultPort = "9020"
+)
+
+//go:generate mockgen -source=router.go -package mocks -destination=mocks/router.go
+type Router interface {
+	Submit(
+		ctx context.Context,
+		processedTxs txpool.TxMap,
+		namespace string,
+		epoch,
+		unixTimestamp uint64,
+	) error
+
+	QueryTransactions(ctx context.Context, req *shardtypes.QueryTransactionsRequest) (
+		*shardtypes.QueryTransactionsResponse,
+		error,
+	)
+
+	Shutdown()
+	Run() error
+}
+
+var _ routerv1.MsgServer = (*routerImpl)(nil)
+var _ Router = (*routerImpl)(nil)
+
+type routerImpl struct {
+	routerv1.MsgServer
+
+	provider       Provider
+	ShardSequencer shard.TransactionHandlerClient
+	ShardQuerier   shardtypes.QueryClient
+
+	port     string
+	shutdown func()
+}
+
+func (r *routerImpl) Shutdown() {
+	if r.shutdown != nil {
+		r.shutdown()
+	}
+}
+
+func (r *routerImpl) Run() error {
+	server := grpc.NewServer()
+	routerv1.RegisterMsgServer(server, r)
+	listener, err := net.Listen("tcp", ":"+r.port)
+	if err != nil {
+		return eris.Wrapf(err, "error listening to port %s", r.port)
+	}
+	go func() {
+		err = eris.Wrap(server.Serve(listener), "error serving server")
+		if err != nil {
+			zerolog.Fatal().Err(err).Msg(eris.ToString(err, true))
+		}
+	}()
+	r.shutdown = server.GracefulStop
+	return nil
+}
+
+func New(sequencerAddr, baseShardQueryAddr string, provider Provider) (Router, error) {
+	rtr := &routerImpl{port: defaultPort, provider: provider}
+
+	conn, err := grpc.Dial(sequencerAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, eris.Wrapf(err, "error dialing shard seqeuncer address at %q", sequencerAddr)
+	}
+	rtr.ShardSequencer = shard.NewTransactionHandlerClient(conn)
+
+	// we don't need secure comms for this connection, cause we're just querying cosmos public RPC endpoints.
+	conn2, err := grpc.Dial(baseShardQueryAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, eris.Wrapf(err, "error dialing evm base shard address at %q", baseShardQueryAddr)
+	}
+	rtr.ShardQuerier = shardtypes.NewQueryClient(conn2)
+	return rtr, nil
+}
+
+const (
+	CodeSuccess = uint32(iota)
+	CodeTxFailed
+	CodeNoResult
+	CodeServerUnresponsive
+	CodeUnauthorized
+	CodeUnsupportedMessage
+	CodeInvalidFormat
+)
+
+// SendMessage is the server impl that receives SendMessage requests from the base shard client.
+func (r *routerImpl) SendMessage(_ context.Context, req *routerv1.SendMessageRequest,
+) (*routerv1.SendMessageResponse, error) {
+	// first we check if we can extract the transaction associated with the id
+	msgType, exists := r.provider.GetMessageByName(req.MessageId)
```

Comment:
i dont really mind either way, but it's probably better to just get a list of messages in New() and create an index/map that stores it in the router struct. Talking through an interface is slow (Go doesn't have zero cost abstraction) and you are copying this object again and again for every message that comes in here.

Reply:
I don't think this is copying objects? I either get the message from my own index, or i get it from Engine's. The tradeoff here has to do with whether we want to store the index twice (both engine and router store) or store it once, and get it through the provider interface.

i don't think talking through an interface to get the message is going to be a significant enough performance hit.

Reply:
`r.provider.GetMessageByName()` returns a copy of the message object by value instead of a reference, so the object `msgType` is copied from engine and deallocated once the function terminates. 

Object allocation is expensive in Go and SendMessage is a hot function (it will be called a lot) in contrast to `New()` that is only called once.

Reply:
the message is an interface. interfaces are always reference/pointer types, not values.

i'm not sure how getting the message interface from engine or from a mapping in `router` changes object allocation. an object is allocated for the message type no matter where i get it from